File: solutions_2692487_1/java/bandesz80/Motes.java
package com.google.codejam.ka;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.PrintWriter;
import java.util.Arrays;
import java.util.Scanner;

public class Motes {

    private static int currMin = Integer.MAX_VALUE;

    public static void main(String args[]) throws FileNotFoundException {
        Scanner scanner = new Scanner(new File(args[0]));
        PrintWriter writer = new PrintWriter(new FileOutputStream("motes.out"));
        try {
            int caseCnt = scanner.nextInt();
            for (int i = 0; i < caseCnt; i++) {
                int mote = scanner.nextInt();
                int moteCnt = scanner.nextInt();
                int motes[] = new int[moteCnt];
                for (int j = 0; j < moteCnt; j++) {
                    motes[j] = scanner.nextInt();
                }
                Arrays.sort(motes);
                currMin = Integer.MAX_VALUE;
                int moves = solve(mote, motes, 0, 0);
                // solve
                writer.println(String.format("Case #%d: %d", i + 1 , moves));
            }
        } finally {
            writer.close();
        }
    }

    public static int solve(int mote, int motes[], int cnt, int idx) {
        if (cnt >= currMin) {
            return cnt;
        }
        int currMote = mote;
        int i = idx;

        for (; i < motes.length && motes[i] < currMote; i++) {
            currMote += motes[i];
        }
        if (i == motes.length) {
            currMin = cnt;
            return cnt;
        } else {
            int withAdding = (currMote == 1 ? Integer.MAX_VALUE : solve(currMote + currMote - 1, motes, cnt + 1, i));
            int withDeleting = solve(currMote, motes, cnt + 1, i + 1);
            return Math.min(withAdding, withDeleting);
        }
    }
}

File: solutions_2692487_1/C++/sebii/motes.cpp
#include <algorithm>
#include <iostream>
#include <set>
#include <vector>

using namespace std;

long long solve(vector<long long> motes, long long A, long long N)
{
    long long j = 0;
    while (j < N && motes[j] < A) {
        A += motes[j];
        ++j;
    }

    long long ans = N - j;
    if (A - 1 == 0)
        return ans;

    long long ans2 = 0;
    while (j < N) {
        A += A - 1;
        ++ans2;
        while (j < N && motes[j] < A) {
            A += motes[j];
            ++j;
        }
        ans = min(ans, ans2 + N - j);
    }

    return ans;
}

int main()
{
    int T;
    cin >> T;

    int id = 1;
    while (T--) {
        long long A, N;
        cin >> A >> N;

        vector<long long> motes(N);
        for (int i = 0; i < N; ++i)
            cin >> motes[i];
        sort(begin(motes), end(motes));

        cout << "Case #" << id << ": " << solve(motes, A, N) << "\n";
        ++id;
    }

    return 0;
}
